compiler/noirc_evaluator/src/ssa/opt/unrolling.rs

```diff
@@ -173,73 +201,511 @@ impl Loops {
+        &self,
+        function: &Function,
+        refs: &HashSet<ValueId>,
+    ) -> (usize, usize) {
+        let mut loads = 0;
+        let mut stores = 0;
+        for block in &self.blocks {
+            for instruction in function.dfg[*block].instructions() {
+                match &function.dfg[*instruction] {
+                    Instruction::Load { address } if refs.contains(address) => {
+                        loads += 1;
+                    }
+                    Instruction::Store { address, .. } if refs.contains(address) => {
+                        stores += 1;
+                    }
+                    _ => {}
+                }
+            }
+        }
+        (loads, stores)
+    }
+
+    /// Count the number of instructions in the loop, including the terminating jumps.
+    fn count_all_instructions(&self, function: &Function) -> usize {
+        self.blocks
+            .iter()
+            .map(|block| {
+                let block = &function.dfg[*block];
+                block.instructions().len() + block.terminator().map(|_| 1).unwrap_or_default()
+            })
+            .sum()
+    }
+
+    /// Count the number of increments to the induction variable.
+    /// It should be one, but it can be duplicated.
+    /// The increment should be in the block where the back-edge was found.
+    fn count_induction_increments(&self, function: &Function) -> usize {
+        let back = &function.dfg[self.back_edge_start];
+        let header = &function.dfg[self.header];
+        let induction_var = header.parameters()[0];
+
+        back.instructions().iter().filter(|instruction|  {
+            let instruction = &function.dfg[**instruction];
+            matches!(instruction, Instruction::Binary(Binary { lhs, operator: BinaryOp::Add, rhs: _ }) if *lhs == induction_var)
+        }).count()
+    }
+
+    /// Decide if this loop is small enough that it can be inlined in a way that the number
+    /// of unrolled instructions times the number of iterations would result in smaller bytecode
+    /// than if we keep the loops with their overheads.
+    fn is_small_loop(&self, function: &Function, cfg: &ControlFlowGraph) -> bool {
+        self.boilerplate_stats(function, cfg).map(|s| s.is_small()).unwrap_or_default()
+    }
+
+    /// Collect boilerplate stats if we can figure out the upper and lower bounds of the loop,
+    /// and the loop doesn't have multiple back-edges from breaks and continues.
+    fn boilerplate_stats(
+        &self,
+        function: &Function,
+        cfg: &ControlFlowGraph,
+    ) -> Option<BoilerplateStats> {
+        let Ok(Some((lower, upper))) = self.get_const_bounds(function, cfg) else {
+            return None;
+        };
+        let Some(lower) = lower.try_to_u64() else {
+            return None;
+        };
+        let Some(upper) = upper.try_to_u64() else {
+            return None;
+        };
+        let Ok(refs) = self.find_pre_header_reference_values(function, cfg) else {
```

Comment:
This method, `count_loads_and_stores`, and `count_all_instructions` all need to iterate over all blocks in the loop.  We could instead iterate over the loop blocks once and have per block methods.

Reply:
This probably isn't a big deal as going through only the blocks again shouldn't have a crazy cost, but just noting it.

Reply:
True, I thought about it as well. On the other hand I thought that 1) having two dedicated methods makes it easier to understand and test, and 2) `count_all_instructions` doesn't need lookups, just the length of the ID vector. Forcing them into a single loop function felt artificial at the time, but at the same time it's unlikely to be used for anything else, so I suppose I could consolidate them into the stats method.